### PR TITLE
[BUG FIX] add fail check when get fail in case wait complete block

### PR DESCRIPTION
**Pull request by mss1213**
### BUG Description
When use wait_complete policy for HiCache prefetch and batch_get meet error, it will occur that the req in prefetch will not be scheduled because the operation will not be completed; This PR would change the operation complete condition, add one flag "failed" for the unexpected error.
image
### Reproduction
use hicache_storage_prefetch_policy = 'wait_complete' and produce some error on batch_get(such as modify its return value)。You can also use offline script， and the program will hang up.

```
import sglang as sgl
from sglang.lang.api import Engine
import json
import os
import time
import random

model_path = "/data01/models/DeepSeek-R1"


def generate_prompts():
    random.seed(100)
    prompts = []
    for i in range(4):
        prompt = []
        for j in range(1000):
            prompt.append(random.randint(0, 16000))
        prompts.append(prompt)
    return prompts


def main():
    prompts = generate_prompts()
    print(f"prompts: {len(prompts)}")

    llm = Engine(
            model_path=model_path,
            enable_hierarchical_cache=True,
            hicache_write_policy="write_through",
            mem_fraction_static=0.3,
            hicache_ratio=1.1,
            hicache_io_backend='kernel',
            hicache_mem_layout='page_first',
            hicache_storage_backend='file',
            hicache_storage_prefetch_policy='wait_complete',
            log_level="info",
            trust_remote_code=True,
            tp_size=8,
            page_size=16,
            enable_torch_compile=False,
            json_model_override_args='{"num_hidden_layers": 10}',
        )

    sampling_params = {"temperature": 0, "max_new_tokens": 20, "ignore_eos": True}
    logprob = True
    extra_params = {}
    if logprob:
        extra_params["return_logprob"] = True
        extra_params["top_logprobs_num"] = 5

    bs = 1
    for i in range(0, len(prompts), bs):
        batch_prompts = prompts[i:i + bs]
        print(f"batch {i // bs}: {len(batch_prompts)}")
        start_time = time.perf_counter()
        outputs = llm.generate(sampling_params=sampling_params, input_ids=batch_prompts, **extra_params)
        end_time = time.perf_counter()
        print(f"outputs: {outputs}")
        print(f"batch {i // bs} cost time: {end_time - start_time:.4f}s")
    print(f"all done")

if __name__ == "__main__":
    main()
```
### Thoughts
I think 'wait_complete' means that scheduler should wait for the operation complete (success or fail both are completed) but not all tokens are prefetched success. Actually batch_get can not always return success even use retry (or use batch_exists first) when use remote storage.

### CODE REVIEW

File: python/sglang/srt/mem_cache/hiradix_cache.py [hunk 1]
```diff
@@ -457,7 +457,13 @@ def can_terminate_prefetch(self, operation: PrefetchOperation):
...
                 group=self.tp_group,
             )
-            can_terminate = bool(can_terminate.item())
+            operation_fail = torch.tensor(operation.failed, dtype=torch.int)
+            torch.distributed.all_reduce(
+                operation_fail,
+                op=torch.distributed.ReduceOp.MAX,
+                group=self.tp_group,
+            )
+            can_terminate = bool(can_terminate.item()) or bool(operation_fail.item())
```
[the remark this answers is not in the document]
Comment (mss1213) [reply python/sglang/srt/mem_cache/hiradix_cache.py]: naming style is consistent with can_terminate

**Comment by xiezhq-hermann** [comment]
@pansicheng I think there might be a better way to handle failure in wait_complete case, can you help take a look?

**Comment by pansicheng** [comment]
> @pansicheng I think there might be a better way to handle failure in wait_complete case, can you help take a look?

Good catch! Handling prefetch failures can also help reduce the waiting time for failed requests in timeout scenarios by allowing inference to be scheduled earlier. Here’s a question I have: when an operation fails in one rank, is it necessary to stop prefetching across all ranks? Perhaps we could implement a more fine-grained control mechanism — for example, only stopping prefetching when there is at least one failed rank, and the total number of prefetched tokens from all non-failed ranks exceeds the minimum number of prefetched tokens among the failed ranks.

**Comment by mss1213** [comment]
> > @pansicheng I think there might be a better way to handle failure in wait_complete case, can you help take a look?
> 
> Good catch! Handling prefetch failures can also help reduce the waiting time for failed requests in timeout scenarios by allowing inference to be scheduled earlier. Here’s a question I have: when an operation fails in one rank, is it necessary to stop prefetching across all ranks? Perhaps we could implement a more fine-grained control mechanism — for example, only stopping prefetching when there is at least one failed rank, and the total number of prefetched tokens from all non-failed ranks exceeds the minimum number of prefetched tokens among the failed ranks.

@pansicheng  Yes, fail fast is a effective mechanism and the sticking point  may be the synchronization timing because of the overhead of all_reduce. Since the prefetch operation use batch_get in fixed batch size，if synchronize all tp after every batch_get in HiCache Controller(when fail, we can break this prefetch operation), it may increase total time of prefetch operation(especially batch_get fail should be low probability event), so i think that there should be a trade-off. This PR is a simple fix, it only check in the scheduler, and will be good in batch_get success(the prefetch thread will not check), but when fail, some rank wouldn't fail fast.

**Comment by xiezhq-hermann** [comment]
@mss1213 thanks again for the PR, I made some changes, including reusing the `mark_done` instead of introducing a new failed flag, and only to do `all_reduce` once. Can you help take a look and do a test if possible to see if it aligns with your goal : )

**Review by xiezhq-hermann — APPROVED** [review]

**Comment by mss1213** [comment]
> @mss1213 thanks again for the PR, I made some changes, including reusing the `mark_done` instead of introducing a new failed flag, and only to do `all_reduce` once. Can you help take a look and do a test if possible to see if it aligns with your goal : )

@xiezhq-hermann Thanks for your refine。I have tested the code and it seems work well.
image
